File: Niveau 3/Niveau 3.6/densité1.hs
--------------------------------------------------------------------------------
-- Name:        Densité du plastique
-- Purpose:     France-IOI niveau 3.6
--
-- Author:      Sébastien
--
-- Created:     24/09/2022
--------------------------------------------------------------------------------
import Data.List
import Control.Monad

-- Recherche une valeur dans une liste par dichotomie
-- Retourne 1 si la valeur est présente dans la liste, 0 sinon
rechercher :: Ord a => a -> [a] -> Int -> Int -> Int
rechercher x xs debut fin =
    let moitie = (fin + debut) `div` 2
        milieu = xs !! moitie
     in if x == milieu then 1
        else if x < milieu && debut < moitie then rechercher x xs debut (moitie-1)
        else if x > milieu &&   fin > moitie then rechercher x xs (moitie+1) fin
        else 0

-- Récupère une valeur sur l'entrée standard et regarde si présente dans une liste        
traiter :: (Read a, Ord a) => [a] -> IO()
traiter xs = do
    x <- readLn
    print $ rechercher x xs 0 (length xs - 1)

main :: IO()
main = do
    n <- readLn :: IO Int
    ligne <- getLine
    let densites = sort (map read (words ligne)) :: [Int]
    q <- readLn :: IO Int
    replicateM_ q $ traiter densites
    

File: Niveau 3/Niveau 3.6/densité1.py
#-------------------------------------------------------------------------------
# Name:        Densité du plastique
# Purpose:     France-IOI niveau 3.6
#
# Author:      Sébastien
#
# Created:     22/09/2022
#-------------------------------------------------------------------------------

import sys

def rechercher(liste, valeur):
    debut = 0
    fin = len(liste) - 1
    while debut <= fin:
        milieu = (debut + fin) // 2
        if valeur > liste[milieu]:
            debut = milieu + 1
        elif valeur < liste[milieu]:
            fin = milieu - 1
        else:
            return 1
    return 0

def main():
    n = int(input())
    densites = list(map(int, sys.stdin.readline().split()))
    densites.sort()

    q = int(input())
    for _ in range(q):
        densite = int(sys.stdin.readline())
        print(rechercher(densites, densite))

if __name__ == '__main__':
    main()

"""
7
1 3 6 7 9 11 20
9
0
1
2
3
10
11
12
20
21
"""